README
NAME

       UHC - Source code filter for UHC script

SYNOPSIS
    In your script:

      use UHC;

    At command prompt:

      perl.exe  yourscript.pl  wild*  *card  and  '*quote*'  on MSWin32
      perl55    yourscript.pl  wild*  *card  and  '*quote*'  on MSWin32
      perl56    yourscript.pl  wild*  *card  and  '*quote*'  on MSWin32
      perl58    yourscript.pl  wild*  *card  and  '*quote*'  on MSWin32
      perl510   yourscript.pl  wild*  *card  and  '*quote*'  on MSWin32
      perl512   yourscript.pl  wild*  *card  and  '*quote*'  on MSWin32
      perl64    yourscript.pl  wild*  *card  and  '*quote*'  on MSWin32

DESCRIPTION

      This software is "JPerl on the Modern Perl" written by Perl.
      We can write the script in UHC (Unified Hangul Code) again.

      This software treats UHC directly. Therefore, there is not UTF8 flag.

INSTALLATION BY MAKE (for UNIX)

To install this software by make, type the following:

   perl Makefile.PL
   make
   make test
   make install

   Rename and install strict.pm_ of this distribution to strict.pm if your system
   doesn't have strict.pm.

INSTALLATION WITHOUT MAKE (for DOS like system)

To install this software without make, type the following:

   perl pMakefile.PL    --- pMakefile.PL makes "pmake.bat" only, and ...
   pmake.bat
   pmake.bat test

   pmake.bat install           --- install to current using Perl
   perl55   pmake.bat install  --- install to Perl5.5
   perl56   pmake.bat install  --- install to Perl5.6
   perl58   pmake.bat install  --- install to Perl5.8
   perl510  pmake.bat install  --- install to Perl5.10
   perl512  pmake.bat install  --- install to Perl5.12
   perl64   pmake.bat install  --- install to Perl64

   Rename and install strict.pm_ of this distribution to strict.pm if your system
   doesn't have strict.pm.

   pmake.bat dist       --- make distribution package
   pmake.bat ptar.bat   --- make perl script "ptar.bat"

PROGRAM COMPOSITION

   UHC.pm                --- source code filter for UHC script
   Euhc.pm               --- run-time routines for UHC.pm
   perl55.bat            --- find and run perl5.5  without %PATH% settings
   perl56.bat            --- find and run perl5.6  without %PATH% settings
   perl58.bat            --- find and run perl5.8  without %PATH% settings
   perl510.bat           --- find and run perl5.10 without %PATH% settings
   perl512.bat           --- find and run perl5.12 without %PATH% settings
   perl64.bat            --- find and run perl64   without %PATH% settings
   strict.pm_            --- dummy strict.pm
   warnings.pm_          --- poor warnings.pm
   warnings/register.pm_ --- poor warnings/register.pm

DEPENDENCIES

This software requires perl5.00503 or later.

LICENSE AND COPYRIGHT

This software is free software; you can redistribute it and/or
modify it under the same terms as Perl itself. See perlartistic.

This software is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

MAINTAINER

This project was originated by INABA Hitoshi <ina@cpan.org>.

"UHC" IN THIS SOFTWARE

 The "UHC" in this software means widely codeset than general UHC.
This software use two algorithms to handle UHC.

* ALGORITHM #1

 When the character is taken out of the octet string, it is necessary to
distinguish a single octet character and the double octet character.
The distinction is done only by first octet.

    Single octet code is:
      0x00-0x80 and 0xFF

    Double octet code is:
      First octet   0x81-0xFE
      Second octet  0x00-0xFF (All octet)

    MALFORMED single octet code is:
      0x81-0xFE
      * Final octet of string like first octet of double octet code

See also code table:

         Single octet code

   0 1 2 3 4 5 6 7 8 9 A B C D E F 
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 0|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*| 0x00-0x80
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 1|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 2|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 3|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 4|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 5|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 6|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 7|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 8|*| | | | | | | | | | | | | | | |
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 9| | | | | | | | | | | | | | | | |
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 A| | | | | | | | | | | | | | | | |
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 B| | | | | | | | | | | | | | | | |
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 C| | | | | | | | | | | | | | | | |
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 D| | | | | | | | | | | | | | | | |
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 E| | | | | | | | | | | | | | | | |
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 F| | | | | | | | | | | | | | | |*| 0xFF
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+


                                 Double octet code
            First octet                                     Second octet

   0 1 2 3 4 5 6 7 8 9 A B C D E F                 0 1 2 3 4 5 6 7 8 9 A B C D E F 
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+               +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 0| | | | | | | | | | | | | | | | |              0|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*| 0x00-0xFF
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+               +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 1| | | | | | | | | | | | | | | | |              1|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+               +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 2| | | | | | | | | | | | | | | | |              2|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+               +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 3| | | | | | | | | | | | | | | | |              3|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+               +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 4| | | | | | | | | | | | | | | | |              4|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+               +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 5| | | | | | | | | | | | | | | | |              5|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+               +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 6| | | | | | | | | | | | | | | | |              6|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+               +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 7| | | | | | | | | | | | | | | | |              7|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+               +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 8| |*|*|*|*|*|*|*|*|*|*|*|*|*|*|*| 0x81-0xFE    8|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+               +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 9|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|              9|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+               +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 A|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|              A|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+               +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 B|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|              B|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+               +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 C|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|              C|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+               +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 D|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|              D|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+               +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 E|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|              E|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+               +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 F|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*| |              F|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+               +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+


    *MALFORMED* Single octet code

Final octet of string like first octet of double octet code

Even if malformed, it is not ignored and not deleted automatically.
For example, Euhc::chop function returns this octet.

   0 1 2 3 4 5 6 7 8 9 A B C D E F 
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 0| | | | | | | | | | | | | | | | |
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 1| | | | | | | | | | | | | | | | |
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 2| | | | | | | | | | | | | | | | |
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 3| | | | | | | | | | | | | | | | |
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 4| | | | | | | | | | | | | | | | |
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 5| | | | | | | | | | | | | | | | |
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 6| | | | | | | | | | | | | | | | |
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 7| | | | | | | | | | | | | | | | |
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 8| |M|M|M|M|M|M|M|M|M|M|M|M|M|M|M| 0x81-0xFE
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 9|M|M|M|M|M|M|M|M|M|M|M|M|M|M|M|M|
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 A|M|M|M|M|M|M|M|M|M|M|M|M|M|M|M|M|
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 B|M|M|M|M|M|M|M|M|M|M|M|M|M|M|M|M|
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 C|M|M|M|M|M|M|M|M|M|M|M|M|M|M|M|M|
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 D|M|M|M|M|M|M|M|M|M|M|M|M|M|M|M|M|
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 E|M|M|M|M|M|M|M|M|M|M|M|M|M|M|M|M|
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 F|M|M|M|M|M|M|M|M|M|M|M|M|M|M|M| |
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+

* ALGORITHM #2

Against algorithm.1, when the range of the character by tr/// is specified, only the
following character codes are effective.

    Single octet code is:
      0x00-0x80 and 0xFF

    Double octet code is:
      First octet   0x81-0xFE
      Second octet  0x41-0x5A, 0x61-0x7A and 0x81-0xFE

For instance, [\x81\x00-\x82\xFF] in script means [\x81\x82][\x41-\x5A\x61-\x7A\x81-\xFE].

See also code table:

         Single octet code

   0 1 2 3 4 5 6 7 8 9 A B C D E F 
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 0|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*| 0x00-0x80
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 1|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 2|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 3|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 4|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 5|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 6|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 7|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|*|
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 8|*| | | | | | | | | | | | | | | |
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 9| | | | | | | | | | | | | | | | |
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 A| | | | | | | | | | | | | | | | |
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 B| | | | | | | | | | | | | | | | |
  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+

Changes
Revision history for this software.

0.71  2011/01/26 00:00:00
	- support @homepage environment of @nifty. (Debugging ended thanks to valuable cooperation of jscripter)
	; created by INABA Hitoshi

0.70  2011/01/14 00:00:00
	- remove multibyte anchoring of UTF2 and OldUTF8.
	- propagate @INC by -I switch from script.pl to script.pl.e.
	- propagate $^W by -w switch from script.pl to script.pl.e.
	- support chdir(\x5c ends path) by perl5.006 or perl5.00800 on MSWin32(jacode.pl library required).
	- support chdir(\x5c ends long path) by perl5.008001 or later, perl5.010 or perl5.012 on MSWin32(Win32 module required).
	- support GBK.
	- bundle strict.pm_ as dummy strict.pm.
	- add perl55.bat and perl56.bat.
	; created by INABA Hitoshi

0.69  2010/12/21 00:00:00
	- support MacOS by Sjis software.
	- correct open mode before flock().
	; beside "escape" by Delight Prescott of MacPerl Power and Ease

0.68  2010/11/09 00:00:00
	- re-correct 244_3open.t and 245_2binmode.t.
	; created by INABA Hitoshi

0.67  2010/11/08 00:00:00
	- correct 244_3open.t and 245_2binmode.t.
	; created by INABA Hitoshi

0.66  2010/11/07 00:00:00
	- emulate two-argument binmode() on perl5.005.
	- emulate three-argument open() on perl5.005.
	- dummy functions of utf8.pm and bytes.pm.
	- correct write locking routine.
	- bundle warnings.pm_ and warnings/register.pm_.
	; created by INABA Hitoshi

0.65  2010/09/16 00:00:00
	- check code page on DOS-like system before test.
	; created by INABA Hitoshi

0.64  2010/09/14 00:00:00
	- support SjisTk module.
	- support \l, \u, lcfirst and ucfirst.
	- add test of Latin1.
	; created by INABA Hitoshi

0.63  2010/07/04 00:00:00
	- add test of Big5Plus, Big5HKSCS, GB18030 and UHC.
	; created by INABA Hitoshi

0.62  2010/07/03 00:00:00
	- redefine \b and \B.
	; created by INABA Hitoshi

0.61  2010/06/27 00:00:00
	- fix OldUTF8.
	- initialize tr variables on demand.
	- fix Esjis::lc and Esjis::uc's bug.
	; created by INABA Hitoshi

0.60  2010/06/16 00:00:00
	- back *.pm files from /lib.
	; created by INABA Hitoshi

0.59  2010/06/14 00:00:00
	- move *.pm files to /lib.
	- fix Makefile.PL.
	- fix m/[$@]/ to m/[\$\@]/.
	- add HP15 and INFORMIXV6ALS test.
	- support Big5-HKSCS.
	; created by INABA Hitoshi

0.58  2010/06/07 00:00:00
	- fix escapeshellcmd.
	; created by INABA Hitoshi

0.57  2010/06/02 00:00:00
	- do BEGIN { unshift @INC, '/Perl/site/lib/Sjis'; } in Esjis.pm.
	- support 'no Module'.
	- remove special routine of Tk module.
	- test does only one version of perl.
	- support TAP(Test Anything Protocol).
	- add ShiftJIS::Regexp's test.
	- rename Makefile.PL to pMakefile.PL(Poor Makefile.PL) for DOS like system.
	- standard Makefile.PL for UNIX.
	- warn if chdir "\x5c" ended path on $^W is true.
	; created by INABA Hitoshi

0.56  2010/04/21 00:00:00
	- add JPerl's test.
	- fix [^]] in regexp from <'^' then ']'> to <not ']'>.
	- [] in regexp make die "unmatched [] in regexp at ...".
	- support \b and \B in regexp.
	; created by INABA Hitoshi

0.55  2010/04/16 00:00:00
	- test.pl support perl5.012.
	- support when as a statement modifier.
	- support := operator.
	; created by INABA Hitoshi

0.54  2010/04/02 00:00:00
	- fix /i modifier with character class '[ccc]' in regexp.
	- support old UTF-8.
	; created by INABA Hitoshi

0.53  2010/04/01 00:00:00
	- support Latin-1.
	; created by INABA Hitoshi

0.52  2010/03/17 00:00:00
	- bugfix ${ foo } in regexp.
	- remove modules (re.pm, Fcntl.pm and Symbol.pm).
	- call flock in Esjis::do and Esjis::require.
	- add perl64.bat.
	- Makefile.pl support UNIX like system.
	; created by INABA Hitoshi

0.51  2010/03/07 00:00:00
	- remove ** UNAUTHORIZED RELEASE ** modules (Fcntl.pm and Symbol.pm).
	; created by INABA Hitoshi

0.50  2010/03/07 00:00:00
	- correct INFOMIXV6ALS to INFORMIXV6ALS (sorry!! forgive me!! ;X).
	- add UNIX shell escape.
	- glob @ARGV when [?*] included only.
	- avoid warning in Esjis::e.
	- flock on write and read script.
	- add substitute modules (Fcntl.pm and Symbol.pm).
	- remove modules (Carp.pm and FindBin.pm).
	; created by INABA Hitoshi

0.49  2010/02/10 00:00:00
	- add perl512.bat
	- redefined character class \d \s \w for perl 5.11 or later.
	- fix test.pl and comment of BUGS AND LIMITATIONS about /o modifier.
	; created by INABA Hitoshi

0.48  2010/01/28 00:00:00
	- Call Module->import() when Module->can('import') is true only.
	- No escape when require jcode.pl and jacode.pl.
	- Remove limitation of two or more here-documents in a line.
	; created by INABA Hitoshi

0.47  2010/01/02 00:00:00
	- shorten script of EUCJP.pm and UTF2.pm.
	- glob always function as File::DosGlob::glob.
	- ignore use module File::DosGlob, Wildcard, Japanese and JExt.
	; created by INABA Hitoshi

0.46  2009/12/02 00:00:00
	- fix bug of split q//.
	; created by INABA Hitoshi

0.45  2009/11/30 00:00:00
	- fix bug of Esjis::split when string is empty.
	; created by INABA Hitoshi

0.44  2009/11/16 00:00:00
	- support given and when keywords.
	- support \gN, \g{N} and \g{-N} in s///.
	- support stacked filetest operators.
	; created by INABA Hitoshi

0.43  2009/10/12 00:00:00
	- support UTF-2 (aka UTF-8).
	- fix bug of join separated multiple octet of EUCJP, INFOMIXV6ALS and GB18030.
	; created by INABA Hitoshi

0.42  2009/09/28 00:00:00
	- rewrite split(m/^/) to split(m/^/m) on any version of perl.
	- fix bug of function Sjis::substr().
	; created by INABA Hitoshi

0.41  2009/09/18 00:00:00

pMakefile.PL
######################################################################
#
# pMakefile.pl - Poor Makefile.pl for Sjis software
#
#                  http://search.cpan.org/dist/Sjis/
#
# Copyright (c) 2008, 2009, 2010 INABA Hitoshi <ina@cpan.org>
######################################################################

use strict;

# make pmake.bat
if ($^O =~ /\A (?: MSWin32 | NetWare | symbian | dos ) \z/oxms) {
    open(FH_MAKEBAT, '>pmake.bat') || die "Can't open file: pmake.bat\n";
    print FH_MAKEBAT <<'END';
@rem = '--*-Perl-*--
@echo off
if "%OS%" == "Windows_NT" goto WinNT
perl -x -S "%0" %1 %2 %3 %4 %5 %6 %7 %8 %9
goto endofperl
:WinNT
perl -x -S "%0" %*
if NOT "%COMSPEC%" == "%SystemRoot%\system32\cmd.exe" goto endofperl
if %errorlevel% == 9009 echo You do not have Perl in your PATH.
goto endofperl
@rem ';
#!perl
#line 14
undef @rem;
END
    print FH_MAKEBAT <DATA>;
    close FH_MAKEBAT;
}

# make pmake
else {
    open(FH_MAKEBAT, '>pmake') || die "Can't open file: pmake\n";
    print FH_MAKEBAT <<END;
#!$^X
END
    print FH_MAKEBAT <DATA>;
    close FH_MAKEBAT;
    chmod 0755, 'pmake';
}

__END__
######################################################################
#
# pmake - Poor make for Sjis software
#
#                  http://search.cpan.org/dist/Sjis/
#
# Copyright (c) 2008, 2009, 2010 INABA Hitoshi <ina@cpan.org>
######################################################################

use strict;
use File::Path;
use File::Copy;
use File::Basename;
use Test::Harness;
use Archive::Tar;
use Compress::Zlib;

unless (@ARGV) {
    if ($^O =~ /\A (?: MSWin32 | NetWare | symbian | dos ) \z/oxms) {
        die <<END;

usage: pmake
usage: pmake test
usage: pmake install
usage: pmake dist
usage: pmake ptar.bat

END
    }
    else {
        die <<END;

usage: pmake
usage: pmake test
usage: pmake install
usage: pmake dist
usage: pmake ptar

END
    }
}

# get file list
open(FH_MANIFEST, 'MANIFEST');
chomp(my @file = <FH_MANIFEST>);
close FH_MANIFEST;

for my $target (@ARGV) {
    if ($target eq 'test') {
        my @test = grep m{ \A (?: test\.pl | t/.+\.t ) \z }xmsi, @file;
        runtests(@test);
    }
    elsif ($target eq 'install') {

        # install *.pm files to /Perl/site/lib
        my($perl_site_lib) = grep(m{/site/}xmsi, @INC);
        for (grep m/ \. pm \z /xmsi, @file) {
            if (m#^(.+)/#) {
                print STDERR "copy $_ $perl_site_lib/$1\n";
                mkpath("$perl_site_lib/$1", 0, 0777) unless -d "$perl_site_lib/$1";
                copy($_, "$perl_site_lib/$1");
            }
            else {
                print STDERR "copy $_ $perl_site_lib\n";
                copy($_, $perl_site_lib);
            }
        }

        # install *.pl, *.bat, *.exe and *.com files to /Perl/bin
        my($perl_bin) = $^X =~ /^(.*)\\[^\\]*$/;
        for (grep m/ \. (?: pl | bat | exe | com ) \z /xmsi, @file) {
            next if m/(?: Makefile | pMakefile | test ) \.pl  $/xmsi;
            next if m/(?: pmake | ptar )                \.bat $/xmsi;
            if (m#^(.+)/#) {
                print STDERR "copy $_ $perl_bin/$1\n";
                mkpath("$perl_site_lib/$1", 0, 0777) unless -d "$perl_site_lib/$1";
                copy($_, "$perl_bin/$1");
            }
            else {
                print STDERR "copy $_ $perl_bin\n";
                copy($_, $perl_bin);
            }
        }
    }
    elsif ($target eq 'dist') {

        # make work directory
        my($changes) = grep(/^Changes$/i,@file);
        open(FH_CHANGES, $changes) || die "Can't open file: $changes\n";
        my $version = 0;
        while (<FH_CHANGES>) {
            if (/^(\d+\.\d+(\.\d+\.\d+)*)/) {
                if ($1 > $version) {
                    $version = $1;
                }
            }
        }
        close FH_CHANGES;
        my $dirname  = dirname($file[0]);
        my $basename = basename($file[0], '.pm','.pl','.bat');
        my $tardir;
        if ($dirname eq 'lib') {
            $tardir = "$basename-$version";
        }
        elsif ($dirname ne '.') {
            $dirname =~ tr#/#-#;
            $tardir = "$dirname-$basename-$version";
        }
        else {
            $tardir = "$basename-$version";
        }
        rmtree($tardir,0,0);

        # rewrite META.yml file
        if(my($metayml) = grep(/^META\.yml$/i,@file)) {
            if (open(FH_METAYML, $metayml)) {
                $_ = join('',<FH_METAYML>);
                close FH_METAYML;
                if (open(FH_METAYML, ">$metayml")) {
                    s/^(version:\s*)\d+\.\d+(\.\d+\.\d+)*/$1$version/m;
                    binmode FH_METAYML;
                    print FH_METAYML $_;
                    close FH_METAYML;
                }
            }
        }

        # make *.tar file
        my $tar = Archive::Tar->new;
        for my $file (@file) {
            if (-e $file) {
                mkpath(dirname("$tardir/$file"), 0, 0777);
                print STDERR "copy $file $tardir/$file\n";
                copy($file, "$tardir/$file");

#-----------------------------------------------------------------------------
# Sunday December 21, 2008 07:38 PM 
# Fixing world writable files in tarball before upload to CPAN [ #38127 ]
# http://use.perl.org/~bart/journal/38127
#-----------------------------------------------------------------------------
#               $tar->add_files("$tardir/$file");
#-----------------------------------------------------------------------------
                open(FH, $file) || die "Can't open file: $file\n";
                binmode FH;
                local $/ = undef; # slurp mode
                my $data = <FH>;
                close FH;
                if ($file =~ m/\. (com|exe|bat|pl) \z/oxmsi) {
                    $tar->add_data("$tardir/$file", $data, {'mode' => 0775});
                }
                else {
                    $tar->add_data("$tardir/$file", $data, {'mode' => 0664});
                }
#-----------------------------------------------------------------------------
            }
            else {
                die "file: $file is not exists.\n";
            }
        }

        my $tarfile = "$tardir.tar";
        $tar->write($tarfile);
        rmtree($tardir,0,0);

        # make *.tar.gz file
        my $gz = gzopen("$tarfile.gz", 'wb');
        open(FH_TAR, $tarfile) || die "Can't open file: $tarfile\n";
        binmode FH_TAR;
        $gz->gzwrite(join('',<FH_TAR>));
        close FH_TAR;
        $gz->gzclose;
        unlink $tarfile;

        # P.565 Cleaning Up Your Environment
        # in Chapter 23: Security
        # of ISBN 0-596-00027-8 Programming Perl Third Edition.

        # local $ENV{'PATH'} = '.';
        local @ENV{qw(IFS CDPATH ENV BASH_ENV)};

        # untar test
        if ($^O =~ /\A (?: MSWin32 | NetWare | symbian | dos ) \z/oxms) {
            system("pmake.bat ptar.bat");
            system("ptar.bat xzvf $tarfile.gz");
        }
        else {
            system("./pmake ptar");
            system("./ptar xzvf $tarfile.gz");
        }
    }
    elsif ($target =~ /^ptar(?:\.bat)?$/) {

        # make ptar.bat
        if ($^O =~ /\A (?: MSWin32 | NetWare | symbian | dos ) \z/oxms) {
            open(FH_TARBAT, '>ptar.bat') || die "Can't open file: ptar.bat\n";
            print FH_TARBAT <<'END';
@rem = '--*-Perl-*--
@echo off
if "%OS%" == "Windows_NT" goto WinNT
perl -x -S "%0" %1 %2 %3 %4 %5 %6 %7 %8 %9
goto endofperl
:WinNT
perl -x -S "%0" %*
if NOT "%COMSPEC%" == "%SystemRoot%\system32\cmd.exe" goto endofperl
if %errorlevel% == 9009 echo You do not have Perl in your PATH.
goto endofperl
@rem ';
#!perl
#line 14
undef @rem;
END
            print FH_TARBAT <DATA>;
            close FH_TARBAT;
        }

        # make ptar
        else {
            open(FH_TARBAT, '>ptar') || die "Can't open file: ptar\n";
            print FH_TARBAT <<END;
#!$^X
END
            print FH_TARBAT <DATA>;
            close FH_TARBAT;
            chmod 0755, 'ptar';
        }
    }
    else {
        warn "unknown target: $target.\n";
    }
}

__END__
######################################################################
#
# ptar - Poor tar for Sjis software
#
#                  http://search.cpan.org/dist/Sjis/
#
# Copyright (c) 2008, 2009, 2010 INABA Hitoshi <ina@cpan.org>
######################################################################

use strict;
use Compress::Zlib;
use Archive::Tar;

if ($ARGV[0] ne 'xzvf') {
    die <<END;

usage: ptar xzvf file1.tar.gz file2.tar.gz ...

** This ptar supports xzvf option only. **
END
}

for my $gzfile (grep m/\.tar\.gz$/xmsi, @ARGV) {

    my $gz = gzopen($gzfile, 'rb');
    (my $tarfile = $gzfile) =~ s/\.gz$//xmsi;
    open(FH_TAR, ">$tarfile") || die "Can't open file: $tarfile\n";
    binmode FH_TAR;
    while ($gz->gzreadline(my $line)) {
        print FH_TAR $line;
    }
    $gz->gzclose;
    close FH_TAR;

    my $tar = Archive::Tar->new($tarfile,1);
    for my $file ($tar->list_files){
        if (-e $file) {
            print STDERR "skip $file is already exists.\n";
        }
        else {
            print STDERR "x $file\n";
            $tar->extract($file);
        }
    }
    unlink $tarfile;
}

__END__
:endofperl
